sound/soc/sof/topology.c

```diff
@@ -3316,84 +3301,30 @@ int snd_sof_complete_pipeline(struct device *dev,
 	return 1;
 }
 
-/**
- * sof_set_pipe_widget - Set pipe_widget for a component
- * @sdev: pointer to struct snd_sof_dev
- * @pipe_widget: pointer to struct snd_sof_widget of type snd_soc_dapm_scheduler
- * @swidget: pointer to struct snd_sof_widget that has the same pipeline ID as @pipe_widget
- *
- * Return: 0 if successful, -EINVAL on error.
- * The function checks if @swidget is associated with any volatile controls. If so, setting
- * the dynamic_pipeline_widget is disallowed.
- */
-static int sof_set_pipe_widget(struct snd_sof_dev *sdev, struct snd_sof_widget *pipe_widget,
-			       struct snd_sof_widget *swidget)
-{
-	struct snd_sof_control *scontrol;
-
-	if (pipe_widget->dynamic_pipeline_widget) {
-		/* dynamic widgets cannot have volatile kcontrols */
-		list_for_each_entry(scontrol, &sdev->kcontrol_list, list)
-			if (scontrol->comp_id == swidget->comp_id &&
-			    (scontrol->access & SNDRV_CTL_ELEM_ACCESS_VOLATILE)) {
-				dev_err(sdev->dev,
-					"error: volatile control found for dynamic widget %s\n",
-					swidget->widget->sname);
-				return -EINVAL;
-			}
-	}
-
-	/* set the pipe_widget and apply the dynamic_pipeline_widget_flag */
-	swidget->pipe_widget = pipe_widget;
-	swidget->dynamic_pipeline_widget = pipe_widget->dynamic_pipeline_widget;
-
-	return 0;
-}
-
 /* completion - called at completion of firmware loading */
 static int sof_complete(struct snd_soc_component *scomp)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_sof_widget *swidget, *comp_swidget;
-	int ret;
 
-	/* set the pipe_widget and apply the dynamic_pipeline_widget_flag */
+	/* set the pipe_widget */
 	list_for_each_entry(swidget, &sdev->widget_list, list) {
 		switch (swidget->id) {
 		case snd_soc_dapm_scheduler:
 			/*
-			 * Apply the dynamic_pipeline_widget flag and set the pipe_widget field
-			 * for all widgets that have the same pipeline ID as the scheduler widget
+			 * Set the pipe_widget field for all widgets that have
+			 * the same pipeline ID as the scheduler widget
 			 */
 			list_for_each_entry_reverse(comp_swidget, &sdev->widget_list, list)
-				if (comp_swidget->pipeline_id == swidget->pipeline_id) {
-					ret = sof_set_pipe_widget(sdev, swidget, comp_swidget);
-					if (ret < 0)
-						return ret;
-				}
+				if (comp_swidget->pipeline_id == swidget->pipeline_id)
+					comp_swidget->pipe_widget = swidget;
 			break;
 		default:
 			break;
 		}
 	}
 
-	/* verify topology components loading including dynamic pipelines */
-	if (sof_core_debug & SOF_DBG_VERIFY_TPLG) {
-		ret = sof_set_up_pipelines(scomp->dev, true);
-		if (ret < 0) {
-			dev_err(sdev->dev, "error: topology verification failed %d\n", ret);
-			return ret;
-		}
-
-		ret = sof_tear_down_pipelines(scomp->dev);
-		if (ret < 0) {
-			dev_err(sdev->dev, "error: topology tear down pipelines failed %d\n", ret);
-			return ret;
-		}
-	}
-
-	/* set up static pipelines */
-	return sof_set_up_pipelines(scomp->dev, false);
+	return 0;
```

Comment:
NAK on this, on-demand pipelines are opt-in. We want pipelines to remain static by default.